pkg/operator/operator.go

```diff
@@ -284,6 +318,19 @@ func (o *Operator) Run(ctx context.Context) error {
 	if err := setupOperatorConfigControllers(o); err != nil {
 		return errors.Wrap(err, "setup rule-evaluator controllers")
 	}
+	if o.targetStatus != nil {
+		go func() {
+			tickerBuilder := func() Ticker {
+				return NewTicker(targetPollInterval)
+			}
+			o.logger.Info("starting target status reporting polling")
+			for {
+				if err := o.targetStatus.Start(ctx, tickerBuilder); err != nil {
```

Comment:
did you try starting this with the [controller pattern](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/5b8121fe5ec7290c65a24657e24285e035a5def7/pkg/operator/operator_config.go#L104-L130) used by the other controllers? Just thinking that using that would simplify all of your informer setup, but I'm assuming you had a reason not to

Reply:
No. Because my logic polls every 10 seconds and controllers are event driven, if I had used controllers, I would have still needed informers as well. I need to fetch the current Kubernetes state on demand and informers are the most performant option. Correct me if I'm wrong.

Reply:
So I actually think there may be a way to greatly reduce the code complexity by effectively polling using the [`source.Channel`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/source#Channel) controller-runtime construct, e.g.

I haven't tested this exactly, but I'm picturing something like this.

```go
// Create target status controller.
func setupTargetStatusControllers(op *Operator) error {
	events := make(chan event.GenericEvent)
	err := ctrl.NewControllerManagedBy(op.manager).
		Named("target-status").
		// Adding &corev1.Pod{fieldSelector: gmp-system} to manager cache
		// so reconciler client has cached collector pods to fetch.
		Watches(&source.Channel{
			Source: events,
		}, &handler.EnqueueRequestForObject{}).
		Complete(newTargetStatusReconciler(op.manager.GetClient(), events, op.opts))
	if err != nil {
		return errors.Wrap(err, "create target status controller")
	}
	return nil
}

// targetStatusReconciler to hold cached client state and source channel.
type targetStatusReconciler struct {
	client client.Client
	ch     chan event.GenericEvent
	opts   Options
}

// newTargetStatusReconciler creates a new targetStatusReconciler and kicks off
// the reconcile loop by sending the first event to the source channel.
func newTargetStatusReconciler(c client.Client, ch chan event.GenericEvent, opts Options) *targetStatusReconciler {
	// Kick off reconcile loop.
	ch <- event.GenericEvent{}
	return &targetStatusReconciler{
		client: c,
		ch:     ch,
		opts:   opts,
	}
}

// Reconcile polls the collector pods, fetches and aggregates target status and upserts
// into each PodMonitoring's Status field.
func (r *targetStatusReconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
	minDelay := time.NewTimer(10 * time.Second)

	// Do the target status polling and updating...
	// Can use r.opts to control parallelism, fetch timeouts, etc...

	// Check if we beat the timer, otherwise wait.
	<-minDelay.C
	// Re-trigger event
	r.ch <- event.GenericEvent{}

	return reconcile.Result{}, nil
}
```